horovod/spark/common/store.py

```diff
@@ -0,0 +1,385 @@
+        raise NotImplementedError()
+
+    def get_test_data_path(self, idx=None):
+        raise NotImplementedError()
+
+    def saving_runs(self):
+        raise NotImplementedError()
+
+    def get_runs_path(self):
+        raise NotImplementedError()
+
+    def get_run_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_logs_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_filename(self):
+        raise NotImplementedError()
+
+    def get_logs_subdir(self):
+        raise NotImplementedError()
+
+    def exists(self, path):
+        raise NotImplementedError()
+
+    def read(self, path):
+        raise NotImplementedError()
+
+    def get_local_output_dir_fn(self, run_id):
+        raise NotImplementedError()
+
+    def sync_fn(self, run_id):
+        raise NotImplementedError()
+
+    def to_remote(self, run_id, dataset_idx):
+        attrs = self._remote_attrs(run_id, dataset_idx)
+
+        class RemoteStore(object):
+            def __init__(self):
+                for name, attr in attrs.items():
+                    setattr(self, name, attr)
+
+        return RemoteStore()
+
+    def _remote_attrs(self, run_id, dataset_idx):
+        return {
+            'train_data_path': self.get_train_data_path(dataset_idx),
+            'val_data_path': self.get_val_data_path(dataset_idx),
+            'test_data_path': self.get_test_data_path(dataset_idx),
+            'saving_runs': self.saving_runs(),
+            'runs_path': self.get_runs_path(),
+            'run_path': self.get_run_path(run_id),
+            'checkpoint_path': self.get_checkpoint_path(run_id),
+            'logs_path': self.get_logs_path(run_id),
+            'checkpoint_filename': self.get_checkpoint_filename(),
+            'logs_subdir': self.get_logs_subdir(),
+            'get_local_output_dir': self.get_local_output_dir_fn(run_id),
+            'sync': self.sync_fn(run_id)
+        }
+
+    @staticmethod
+    def create(prefix_path, *args, **kwargs):
+        if HDFSStore.matches(prefix_path):
+            return HDFSStore(prefix_path, *args, **kwargs)
+        else:
+            return LocalStore(prefix_path, *args, **kwargs)
+
+
+class PrefixStore(Store):
+    def __init__(self, prefix_path, train_path=None, val_path=None, test_path=None, runs_path=None, save_runs=True):
+        self.prefix_path = self.get_full_path(prefix_path)
+        self._train_path = train_path or self._get_path('intermediate_train_data')
+        self._val_path = val_path or self._get_path('intermediate_val_data')
+        self._test_path = test_path or self._get_path('intermediate_test_data')
+        self._runs_path = runs_path or self._get_path('runs')
+        self._save_runs = save_runs
+        super(PrefixStore, self).__init__()
+
+    def exists(self, path):
+        return self.get_filesystem().exists(self.get_localized_path(path))
```

Comment:
Changed usage in the `TrainingCache` to check new store method `is_parquet_dataset`, which will try to create a PyArrow `ParquetDataset` object.  If the parquet files are malformed then an exception will be raised, which will be caught to return False.

Other usages of `exists` can be for files other than Parquet datasets, or where specific exceptions should be raised.

Reply:
Due to the atomic move, there will be no malformed parquet files if writing fails. For `ParquetDataset`, the existing path will be empty (it should ignore the `_temporary` subdirectory), which should translate into an empty parquet file. This is all `is_parquet_dataset` should / will see in that situation.

Reply:
I've tested this with `ParquetDataset`.  If there is a `_temporary` directory underneath the root of the path, then creating the `ParquetDataset` will fail and `is_parquet_dataset` will return `False`.  That's because a subdirectory is only considered to be valid if it defines a partition of the form `/some/path/dataset/split=key`.  Because `_temporary` does not have this form, it is not considered to be a valid Parquet dataset.

Reply:
sounds like pyarrow deviates here from how scala `spark.read.parquet` works, but sounds like it is well tested on your end